internal/model/ruleset.go

```diff
@@ -0,0 +1,139 @@
+// This is necessary because the tfsdk doesn't support embedding structs.
+type RuleSetRuleTF struct {
+	Metric    types.String `tfsdk:"metric"`
+	MatchType types.String `tfsdk:"match_type"`
+
+	Drop       types.Bool     `tfsdk:"drop"`
+	KeepLabels []types.String `tfsdk:"keep_labels"`
+	DropLabels []types.String `tfsdk:"drop_labels"`
+
+	Aggregations []types.String `tfsdk:"aggregations"`
+
+	AggregationInterval types.String `tfsdk:"aggregation_interval"`
+	AggregationDelay    types.String `tfsdk:"aggregation_delay"`
+}
+
+func (r AggregationRule) IsExactMatch() bool {
+	return r.MatchType == "" || r.MatchType == "exact"
+}
+
+func (r RuleSetRuleTF) ToAPIReq() AggregationRule {
+	return AggregationRule{
+		Metric:    r.Metric.ValueString(),
+		MatchType: r.MatchType.ValueString(),
+
+		Drop:       r.Drop.ValueBool(),
+		KeepLabels: toStringSlice(r.KeepLabels),
+		DropLabels: toStringSlice(r.DropLabels),
+
+		Aggregations: toStringSlice(r.Aggregations),
+
+		AggregationInterval: r.AggregationInterval.ValueString(),
+		AggregationDelay:    r.AggregationDelay.ValueString(),
+
+		ManagedBy: managedByTF,
+	}
+}
+
+// AlignUpstreamWithState reorders the rules from upstream to match the state.
+// This must mantain the semantic equality of the rules, but may reorder them as
+// long as all non-exact rules are present in the same order as they were
+// originally.
+func AlignUpstreamWithState(state AggregationRuleSet, upstream AggregationRuleSet) AggregationRuleSet {
+	// If the two rulesets are semantically different, then we can't reorder them
+	// without changing the meaning of the rules.
+	if !semanticallyEqualOrdering(state, upstream) {
+		return upstream
+	}
+
+	output := make(AggregationRuleSet, 0, max(len(state), len(upstream)))
+
+	// we know there can be no duplicates by metric name, so we can use a map
+	upstreamMap := make(map[string]AggregationRule, len(upstream))
+	for _, rule := range upstream {
+		upstreamMap[rule.Metric] = rule
+	}
+
+	// we iterate over the state rules, and if we find a matching rule in the
+	// upstream, we use that one and remove it from the map.
+	for _, rule := range state {
+		if upstreamRule, ok := upstreamMap[rule.Metric]; ok {
+			output = append(output, upstreamRule)
+			delete(upstreamMap, rule.Metric)
+		}
+	}
+
+	// at this point, all rules with an equivalent in the state have been added
+	// to the output. We can now add the remaining rules from the upstream.
+	for _, rule := range upstream {
+		// we check if the rule still exists in the map, as it may have been
+		// removed by the previous loop.
+		if _, ok := upstreamMap[rule.Metric]; ok {
+			output = append(output, rule)
+		}
+	}
+
+	return output
+}
+
+// semanticallyEqualOrdering checks if the ordering of the rules in two
+// AggregationRuleSets is semantically equal. This means that the non-exact rules
+// are in the same order in both sets.
+func semanticallyEqualOrdering(a, b AggregationRuleSet) bool {
```

Comment:
Am I understanding it correctly that this function only checks whether the ordering of the two lists is semantically equal, but it assumes that the elements in both lists are the same sets of elements?

Reply:
Right, there's really only one scenario this is targeting for now: you submit a ruleset with some ordering, and we re-order it before returning, leading to a diff in terraform that doesn't matter. In that scenario the two lists are the same length and same rules, just different content. If there is any more complicated difference, we can just give up and show the diff, since we know there will be a diff regardless.

Maybe in the future we can do better and minimize the diff, but I don't want to block on that too long for now.

Reply:
I'll also note that the whole thing would be trivial if order didn't matter, since I could change from a terraform list to a terraform set, and then terraform would do all the work to find a minimal diff. Unfortunately we can't do that since re-ordering non-exact rules changes the meaning of the ruleset.